### PR TITLE
Support Testnet4 Network

### PR DESCRIPTION
**Pull request by BinChengZhao**
Fixed: https://github.com/rust-bitcoin/rust-bitcoin/issues/2749

1. Adjust `Network` to support `Testnet4`. (Based on https://github.com/rust-bitcoin/rust-bitcoin/issues/2749#issuecomment-2152366608 as an implementation idea.)

2. Handle conflicting procedure macros and declaration macros, since the original macros can't handle the new `Network::Testnet(TestnetVersion)` structure properly, and make internal implementations compatible with previous versions (e.g., `testnet` for `Testnet3`, `testnet4` for `Testnet4`, to minimize user-side effects). 

3. add `Testnet4` related Params, Block. 

4. optimize compatibility test cases.

5. Regarding `fn: genesis_block`, calling `bitcoin_genesis_tx` with any Network type internally gets a reasonable `merkle_root`, but not `Testnet4`. (By comparison, I confirmed that the built-in transaction data provided is correct, and combined with the fixed `merkle_root` of `Testnet4` the correct block hash can be computed, which is not possible if the transaction data is incorrect.) **I'd appreciate it if some developer could guide me on what the problem is, and I'll work on it.**

Of course, all these changes are based on https://github.com/rust-bitcoin/rust-bitcoin/issues/2749 , and the consensus of the discussion, if you have any comments, please leave a message, I will actively improve, and promote the support of `Testnet4`.

### CODE REVIEW

File: bitcoin/src/network.rs
Comment (Kixunil) [inline bitcoin/src/network.rs:67]: Why default? Wouldn't the latest one become the default eventually? Also it misleads users into thinking `Default` is implemented which it (rightly) isn't.

File: bitcoin/src/network.rs
Comment (Kixunil) [inline bitcoin/src/network.rs:247]: Don't make this `pub`, it'll just trigger panics in user code.

File: bitcoin/src/p2p/mod.rs
Comment (Kixunil) [inline bitcoin/src/p2p/mod.rs:271]: Losing this sucks as it prevented us from creating bugs. It'd be great to somehow restore the macro.

File: bitcoin/src/blockdata/constants.rs
Comment (BinChengZhao) [reply bitcoin/src/blockdata/constants.rs]: At the initial commit, my fmt tool failed, resulting in an incorrect indentation that was already handled at the time of the PR commit.

File: bitcoin/src/network.rs
Comment (BinChengZhao) [reply bitcoin/src/network.rs:67]: Thank you very much for your suggestion. I will change this in time

File: bitcoin/src/network.rs
Comment (BinChengZhao) [reply bitcoin/src/network.rs:247]: Yes, I have also tweaked this so that this internal API should not be pub

File: bitcoin/src/p2p/mod.rs
Comment (BinChengZhao) [reply bitcoin/src/p2p/mod.rs:271]: Ok, I will try to reproduce this macro and try to split it into two macros, because the 'Network' variant we matched cannot satisfy the two internal logics at the same time

File: bitcoin/src/network.rs
Comment (Kixunil) [inline bitcoin/src/network.rs]: Please no. `as usize` was a horrible hack and this even exposes it.

**Comment by BinChengZhao** [comment]
- The original declaration macro  `generate_network_magic_conversion ` was split into two patterns,  Because there was a risk of conflict in doing two code generations in one patten.

- Since compound Enum cannot use `as usize` as the index of slice, `From<Network> for usize` is implemented for `Network`; 

- Because `into() -> size` can no longer be executed in const fn, so the signature of a function is adjusted.

File: bitcoin/src/p2p/mod.rs
Comment (Kixunil) [inline bitcoin/src/p2p/mod.rs:288]: The point of the macro wast o **not** duplicate these items but have them together so that calling the macro synchronizes the conversions.

**Review by Kixunil — COMMENTED** [review]
Please squash the commits, some changes are hard to review this way.

**Comment by BinChengZhao** [comment]
> Please squash the commits, some changes are hard to review this way.

I will pay attention to this in squash commit

File: bitcoin/src/network.rs
Comment (BinChengZhao) [reply bitcoin/src/network.rs]: Okay, let me think about how to remove this behavior, because original code need to use Network as a index to match some parameters

File: bitcoin/src/p2p/mod.rs
Comment (BinChengZhao) [reply bitcoin/src/p2p/mod.rs:288]: The reason for duplicate here is that the patten of the original unique macro cannot satisfy us, because the structure of 'Network' has changed, and it cannot match as before, so I split it into two patten in order to keep the macro.

Do you have any good suggestions here,
- Retain macros (will result in some duplicate)?
- Delete macro ?

File: bitcoin/src/network.rs
Comment (Kixunil) [reply bitcoin/src/network.rs]: I'd just use `match` directly.

File: bitcoin/src/p2p/mod.rs
Comment (Kixunil) [reply bitcoin/src/p2p/mod.rs:288]: I think it should still work since the sides are same but maybe you'll have to change things like `ident` to `tt`. If you can't figure it out and nobody wants to try then it's better to delete the macro entirely.

Comment (BinChengZhao) [reply bitcoin/src/p2p/mod.rs:282]: According to your idea, I tried a variety of optimization methods, `tt` was not too smooth, and finally I merged the two branches to avoid 'duplicate'.

File: bitcoin/src/network.rs
Comment (BinChengZhao) [reply bitcoin/src/network.rs]: Yes I agree with you, I removed the `as usize` behavior.

**Review by Kixunil — COMMENTED** [review]
I tried to find the hash problem for a bit but couldn't so far.

File: bitcoin/src/blockdata/constants.rs
Comment (BinChengZhao) [reply bitcoin/src/blockdata/constants.rs]: Thank you. I'll adjust it to the correct name

**Comment by BinChengZhao** [comment]
> I tried to find the hash problem for a bit but couldn't so far.



> I tried to find the hash problem for a bit but couldn't so far.

I've made several attempts so far and I haven't made any progress. 

Now in doubt some of 'script::Builder::push_slice' internal logic, because Testnet4 provides a new 'genesis_msg' : ` 03 / May / 2024 000000000000000000001 ebd58c244970b3aa9d783bb001011fbe8ea8e98e00e ` might push different branches. But I don't quite understand the internal logic at the moment, it's just conjecture.

**Comment by Kixunil** [comment]
I don't think it's that. Perhaps it'll be faster if you debug-serialize the transaction and compare it to the one in Core repo.

Also important note: while that Core PR is not merged **and released** this PR cannot be merged.

**Comment by BinChengZhao** [comment]
> Also important note: while that Core PR is not merged **and released** this PR cannot be merged.

I understand the need to wait for progress on `bitcoin Core`

**Comment by BinChengZhao** [comment]
> I don't think it's that. Perhaps it'll be faster if you debug-serialize the transaction and compare it to the one in Core repo.

Ok I try to debug 

Since I started writing Rust, C++ has almost forgotten, debugging is also a knowledge review :)

**Comment by BinChengZhao** [comment]
> I don't think it's that. Perhaps it'll be faster if you debug-serialize the transaction and compare it to the one in Core repo.
> 
> Also important note: while that Core PR is not merged **and released** this PR cannot be merged.

After a lot of cross-debug with the `bitcoin-Core` repo, I have found the problem. The new testnet4 uses a compressed public key. After I have processed the corresponding logic, **there is no hack**

Thanks for the tip

**Review by Kixunil — COMMENTED** [review]
The code looks good now, I'll be happy to ACK once Core is released.

File: bitcoin/src/network.rs
Comment (xrem) [inline bitcoin/src/network.rs:201]:
```suggestion
            "testnet4" => Testnet(TestnetVersion::V4),
```
bitcoind node will return `testnet4` and serialization will fail, you should use actual value instead of `test4`

image

```bash
curl --location --request POST 'http://127.0.0.1:48332' \
--header 'Authorization: Basic ...snip...' \
--header 'Content-Type: application/json' \
--data-raw '{
    "method": "getblockchaininfo",
    "params": []
}'
```

```
{
    "result": {
        "chain": "testnet4",
        "blocks": 36409,
        "headers": 36409,
        "bestblockhash": "0000000000000058e9f909a8d5a86afc8f58d39b661b02f07433d4cae8b6481d",
        "difficulty": 44472500.68822794,
        "time": 1721977342,
        "mediantime": 1721974212,
        "verificationprogress": 1,
        "initialblockdownload": false,
        "chainwork": "000000000000000000000000000000000000000000000048dcdd24a9fc06ef13",
        "size_on_disk": 284698607,
        "pruned": false,
        "warnings": [
            "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
        ]
    },
    "error": null
}
```

Comment (Kixunil) [reply bitcoin/src/network.rs:201]: Insane, so the Core devs decided to make it as inconsistent as possible?

Comment (xrem) [reply bitcoin/src/network.rs:201]: PR is not yet finalized, but for now chain serialized as `testnet4` string
https://github.com/bitcoin/bitcoin/pull/29775/files/aba504759caa13338aa574a7839d078b989fbf6d#diff-254a3b28822b27e42585d1a3f4f41cc8d7712c911b7e85be78a2ee11564f4bd8R19

File: bitcoin/src/blockdata/constants.rs
Comment (antonilol) [inline bitcoin/src/blockdata/constants.rs:79]: All the elements are the same, you can write this as `[element; length]`, so `[0; 33]` or `[0x00; 33]`.

File: bitcoin/src/network.rs
Comment (antonilol) [inline bitcoin/src/network.rs:237]:
```suggestion
    pub const fn params(self) -> &'static Params {
```

I see no reason to remove the `const` keyword here, the method still works at compile time.

File: bitcoin/src/p2p/mod.rs
Comment (antonilol) [reply bitcoin/src/p2p/mod.rs:282]: `Network::$network:ident` could also be changed to something like `Network::$network:ident $((TestnetVersion::$testnet_version:ident))?`. The advantage of using idents is it is allowed in patterns and expressions.

Comment (antonilol) [reply bitcoin/src/p2p/mod.rs:282]: I copy pasted the original macro and added the TestnetVersion enum from this PR:

<details>
<summary>diff</summary>

```diff
diff --git a/bitcoin/src/p2p/mod.rs b/bitcoin/src/p2p/mod.rs
index 1121b2fb..30bd8cdc 100644
--- a/bitcoin/src/p2p/mod.rs
+++ b/bitcoin/src/p2p/mod.rs
@@ -246,12 +246,12 @@ impl FromStr for Magic {
 }
 
 macro_rules! generate_network_magic_conversion {
-    ($($network_pat:pat => $magic:ident => $network_expr:expr,)*) => {
+    ($(Network::$network:ident$((TestnetVersion::$testnet_version:ident))? => Magic::$magic:ident,)*) => {
         impl From<Network> for Magic {
             fn from(network: Network) -> Magic {
                 match network {
                     $(
-                        $network_pat => Magic::$magic,
+                        Network::$network$((TestnetVersion::$testnet_version))? => Magic::$magic,
                     )*
                 }
             }
@@ -263,7 +263,7 @@ macro_rules! generate_network_magic_conversion {
             fn try_from(magic: Magic) -> Result<Self, Self::Error> {
                 match magic {
                     $(
-                        Magic::$magic => Ok($network_expr),
+                        Magic::$magic => Ok(Network::$network$((TestnetVersion::$testnet_version))?),
                     )*
                     _ => Err(UnknownMagicError(magic)),
                 }
@@ -274,11 +274,11 @@ macro_rules! generate_network_magic_conversion {
 // Generate conversion functions for all known networks.
 // `Network -> Magic` and `Magic -> Network`
 generate_network_magic_conversion! {
-    Network::Bitcoin => BITCOIN => Network::Bitcoin,
-    Network::Testnet(TestnetVersion::V3) => TESTNET => Network::Testnet(TestnetVersion::V3),
-    Network::Testnet(TestnetVersion::V4) => TESTNET4 => Network::Testnet(TestnetVersion::V4),
-    Network::Signet => SIGNET => Network::Signet,
-    Network::Regtest => REGTEST => Network::Regtest,
+    Network::Bitcoin => Magic::BITCOIN,
+    Network::Testnet(TestnetVersion::V3) => Magic::TESTNET,
+    Network::Testnet(TestnetVersion::V4) => Magic::TESTNET4,
+    Network::Signet => Magic::SIGNET,
+    Network::Regtest => Magic::REGTEST,
 }
 
 impl fmt::Display for Magic {
```

</details>

File: bitcoin/src/blockdata/constants.rs
Comment (antonilol) [inline bitcoin/src/blockdata/constants.rs:132]: `as_ref()` can be a user implemented function, so only calling it once and storing the result in a local variable prevents people from getting `txdata` of one network and the rest of another network.

File: bitcoin/src/blockdata/constants.rs
Comment (antonilol) [inline bitcoin/src/blockdata/constants.rs:82]: No `impl AsRef<Params>` needed here, `&Params` will achieve the same. This is a private function and only called with `Params` directly (not a non-`Params` type that needs AsRef).

**Comment by antonilol** [comment]
Just came across `bitcoin/src/network.rs` lines 275 and 283, testnet4 is missing there in a string, maybe also in other places in the rest of the code.

File: bitcoin/src/network.rs
Comment (antonilol) [inline bitcoin/src/network.rs:339]: This can be made a separate (private?) method and used here (Display) and in the Serialize impl, and the same for reusing FromStr logic in Deserialize.

File: bitcoin/src/network.rs
Comment (BinChengZhao) [reply bitcoin/src/network.rs]: Dear friend, thank you for your prompt. I have modified it according to your suggestion

File: bitcoin/src/blockdata/constants.rs
Comment (BinChengZhao) [reply bitcoin/src/blockdata/constants.rs]: Dear friend, thank you for your clear advice. I have revised it according to your suggestion

File: bitcoin/src/blockdata/constants.rs
Comment (BinChengZhao) [reply bitcoin/src/blockdata/constants.rs]: Dear friend, thank you I have revised it according to your suggestion

**Comment by BinChengZhao** [comment]
> Just came across `bitcoin/src/network.rs` lines 275 and 283, testnet4 is missing there in a string, maybe also in other places in the rest of the code.

Dear friend, I have supplemented the relevant comments according to your tips, such as Hrp notes and Params notes

File: bitcoin/src/network.rs
Comment (BinChengZhao) [reply bitcoin/src/network.rs:339]: Dear friend, I have looked at the code, although the semantics of match is similar, serialization also relies on the data structure of serde, and it seems that it cannot be abstracted into a function

File: bitcoin/src/p2p/mod.rs
Comment (BinChengZhao) [reply bitcoin/src/p2p/mod.rs:281]: > ```rust
> macro_rules! generate_network_magic_conversion {
>     ($(Network::$network:ident$((TestnetVersion::$testnet_version:ident))? => Magic::$magic:ident,)*) => {
>         impl From<Network> for Magic {
>             fn from(network: Network) -> Magic {
>                 match network {
>                     $(
>                         Network::$network$((TestnetVersion::$testnet_version))? => Magic::$magic,
>                     )*
>                 }
>             }
>         }
> 
>         impl TryFrom<Magic> for Network {
>             type Error = UnknownMagicError;
> 
>             fn try_from(magic: Magic) -> Result<Self, Self::Error> {
>                 match magic {
>                     $(
>                         Magic::$magic => Ok(Network::$network$((TestnetVersion::$testnet_version))?),
>                     )*
>                     _ => Err(UnknownMagicError(magic)),
>                 }
>             }
>         }
>     };
> }
> // Generate conversion functions for all known networks.
> // `Network -> Magic` and `Magic -> Network`
> generate_network_magic_conversion! {
>     Network::Bitcoin => Magic::BITCOIN,
>     Network::Testnet(TestnetVersion::V3) => Magic::TESTNET,
>     Network::Testnet(TestnetVersion::V4) => Magic::TESTNET4,
>     Network::Signet => Magic::SIGNET,
>     Network::Regtest => Magic::REGTEST,
> }
> ```

Thank you very much, your macro expression is very exquisite, I borrow your code to modify here

**Comment by Davidson-Souza** [comment]
An afterthought for the API: Since `Network` is now not an unit enum, should `Signet` also represent which signet are we on?

**Comment by Kixunil** [comment]
@Davidson-Souza this was already discussed and the verdict was no. `Netwrok` represents possible "standard" networks supported directly by Bitcoin Core by passing a single switch to it (for `to_core_arg` to be implementable). People who need anything advanced have to use `Params`.

File: bitcoin/src/address/mod.rs
Comment (apoelstra) [inline bitcoin/src/address/mod.rs:196]: In ae178a392f9774f31195aff3b49830e9c00fb3e7:

nit: should have space before `(testnet3)` and a comma after.

File: bitcoin/src/network/mod.rs
Comment (apoelstra) [inline bitcoin/src/network/mod.rs:69]: In ae178a392f9774f31195aff3b49830e9c00fb3e7:

I don't think we should have serde impls on the `TestnetVersion` type. In general I think we should avoid adding serde impls on new types unless we have a clear usecase.

File: bitcoin/src/network/mod.rs
Comment (apoelstra) [inline bitcoin/src/network/mod.rs:121]: In ae178a392f9774f31195aff3b49830e9c00fb3e7:

Why introduce the `from_str_impl` method? Why not just call `from_str` directly?

File: bitcoin/src/network/mod.rs
Comment (Kixunil) [reply bitcoin/src/network/mod.rs:69]: I'd say that `TestnetVersion` is obviously an integer but I agree we should stay conservative.

**Comment by apoelstra** [comment]
In ae178a392f9774f31195aff3b49830e9c00fb3e7:

If possible I'd also like to separate out the commits so that one introduces the `TestnetVersion` enum (with just one variant) and the next one adds the testnet4 variant.

File: bitcoin/src/network/mod.rs
Comment (Kixunil) [reply bitcoin/src/network/mod.rs:121]: It'd make sense if the method accepted `s: impl AsRef<str> + Into<String>` so that `to_owned` is converted to `into` and then also have `visit_string` to save allocation in error case.

**Review by apoelstra — COMMENTED** [review]
Successfully ran local tests on ae178a392f9774f31195aff3b49830e9c00fb3e7.

File: bitcoin/src/network/mod.rs
Comment (BinChengZhao) [reply bitcoin/src/network/mod.rs]: Thank you very much for your valuable advice. It has been adjusted.

**Comment by BinChengZhao** [comment]
> In [ae178a3](https://github.com/rust-bitcoin/rust-bitcoin/commit/ae178a392f9774f31195aff3b49830e9c00fb3e7):
> 
> If possible I'd also like to separate out the commits so that one introduces the `TestnetVersion` enum (with just one variant) and the next one adds the testnet4 variant.

I have split the commits as requested. Each commit is now independent and can be tested or reverted individually :)

File: bitcoin/src/network/mod.rs
Comment (antonilol) [inline bitcoin/src/network/mod.rs:232]: This logic (`Network` to `Params`) is duplicated a few times in the code: this one, `Params::new` and `Params::as_ref`. Not something this pr introduced, but an easy fix. `Params::new` (or any `const fn(Network) -> Params` in general) should be able to cover all needs from the other functions like returning a reference.

**Comment by apoelstra** [comment]
I think we can start to move forward on this. Testnet4 is in the release candidates for Core 28.0 (there are two, possibly a third, and they are targeting a full release next week). The chain also has like 40000 blocks so it is unlikely to see any breaking changes. (Though ofc there was that issue on the mailing list a few weeks ago that led to a 400-block reorg ... but there probably won't be any more such things, and anyway there's no amount of waiting that will let us guarantee it won't happen again.)

utACK 6c97dfccdaa40ffd585dce21dc6a52ee9a30bf65 but I would like followups to:

* Implement the new difficulty/target compuation
* Deprecate ChainHash::TESTNET in favor of a new ChainHash::TESTNET3.

I'm also a little annoyed at the change to `from_array!` to explicitly whitelist the size of the testnet4 genesis message in `Script::push_slice` ... but I don't have a better solution there. Plausibly we should whitelist every integer between 0 and 520. Probably we want to wait for the Rust compiler to support inequalties on const parameters, or this would slow down compile times for no real-world benefit. And meanwhile we should keep this ad-hoc thing.

**Comment by apoelstra** [comment]
Also this needs a rebase. You might want to pull my suggestions into the rebased branch.

**Comment by apoelstra** [comment]
In 207dd82db5f0850613616c55f64f0fd0ce39f578

Can you please keep the old `TESTNET` constant, and just deprecate it?

Also, I am tempted to backport just the `ChainHash` constant changes to 0.32.x.

**Comment by BinChengZhao** [comment]
> Also, I am tempted to backport just the `ChainHash` constant changes to 0.32.

Thanks for your reminding, I have made the changes according to the corresponding suggestions.

**Comment by BinChengZhao** [comment]
> * Implement the new difficulty/target compuation

I'm sorry that I don't have a clear idea for this change. At present, due to time constraints, I cannot continuously respond to PR. I'm very sorry.

**Review by apoelstra — APPROVED** [review]
ACK 3a2f94bc60c82ba8a8e0e608904ccbed0f139445 successfully ran local tests

File: bitcoin/src/network/params.rs
Comment (tcharding) [inline bitcoin/src/network/params.rs:163]: Any reason we don't deprecate this and add `TESTNET3`?

File: bitcoin/src/p2p/mod.rs
Comment (tcharding) [inline bitcoin/src/p2p/mod.rs:279]: Should we deprecate `Magic::TESTNET` also and add `Magic::TESTNET3` ?

**Review by tcharding — APPROVED** [review]
I'm a bit late to get to reviewing this, all my questions can be seen as nits and need not be addressed before merging this.

**Review by tcharding — DISMISSED** [review]
ACK 3a2f94bc60c82ba8a8e0e608904ccbed0f139445

File: bitcoin/src/p2p/mod.rs
Comment (apoelstra) [reply bitcoin/src/p2p/mod.rs:279]: Oh, yeah, good idea.

File: bitcoin/src/network/mod.rs
Comment (apoelstra) [reply bitcoin/src/network/mod.rs:91]: I think we want all serializations to match the command-line argument names in Core.

**Comment by apoelstra** [comment]
Bitcoin Core 28.0 should be out this week, so I'm gonna hold off on merging this til then (not that it really matters since we aren't going to have a release anytime soon).

So plenty of time to address @tcharding's nits if you want @BinChengZhao

File: bitcoin/src/network/mod.rs
Comment (tcharding) [reply bitcoin/src/network/mod.rs:91]: Fair.

**Comment by BinChengZhao** [comment]
> Bitcoin Core 28.0 should be out this week, so I'm gonna hold off on merging this til then (not that it really matters since we aren't going to have a release anytime soon).
> 
> So plenty of time to address @tcharding's nits if you want @BinChengZhao

Ok, thank you for your valuable suggestions. Let me modify the code according to the suggestions :)

**Comment by BinChengZhao** [comment]
> I'm a bit late to get to reviewing this, all my questions can be seen as nits and need not be addressed before merging this.

Thank you very much for your valuable suggestions, I have made the code synchronous changes related to the suggestions

**Review by apoelstra — APPROVED** [review]
ACK 1ef2ab9b62c15602338d5cf97d442805ef614bee successfully ran local tests

File: bitcoin/src/network/mod.rs
Comment (tcharding) [inline bitcoin/src/network/mod.rs:316]: This whole comment is unnecessary, please remove.

**Review by tcharding — COMMENTED** [review]
Thanks for sticking with this, with the comment removed: ACK 1ef2ab9b62c15602338d5cf97d442805ef614bee

**Comment by BinChengZhao** [comment]
> Thanks for sticking with this, with the comment removed: ACK [1ef2ab9](https://github.com/rust-bitcoin/rust-bitcoin/commit/1ef2ab9b62c15602338d5cf97d442805ef614bee)

The corresponding comments have been removed, thank you for your valuable time review

**Review by tcharding — APPROVED** [review]
ACK 2424b654d5bc3e3b509f46b5d261353cdd98ae8b

**Review by apoelstra — APPROVED** [review]
ACK 2424b654d5bc3e3b509f46b5d261353cdd98ae8b successfully ran local tests

**Comment by apoelstra** [comment]
Let's try to backport this.

**Comment by junderw** [comment]
> Let's try to backport this.

Looks like this is going into 0.33?

0.32 and 0.31? How far should we backport?

**Comment by apoelstra** [comment]
Just to 0.32 is fine.

**Comment by tcharding** [comment]
Backported in #3453

**Pull request merged**
